passes/techmap/abc9_ops.cc

```diff
@@ -119,81 +119,127 @@ void mark_scc(RTLIL::Module *module)
 	}
 }
 
-void prep_dff(RTLIL::Module *module)
-{
-	auto design = module->design;
-	log_assert(design);
 
-	SigMap assign_map(module);
-
-	typedef SigSpec clkdomain_t;
-	dict<clkdomain_t, int> clk_to_mergeability;
+void prep_dff_hier(RTLIL::Design *design)
+{
+	pool<IdString> seen;
+	dict<RTLIL::IdString, RTLIL::Selection> selection_vars;
+	auto r YS_ATTRIBUTE(unused) = design->selection_vars.insert(std::make_pair(ID($abc9_flops), RTLIL::Selection(false)));
+	log_assert(r.second);
+	auto r2 YS_ATTRIBUTE(unused) = design->selection_vars.insert(std::make_pair(ID($abc9_cells), RTLIL::Selection(false)));
+	log_assert(r2.second);
+	auto &modules_sel = design->selection_vars.at(ID($abc9_flops));
+	auto &cells_sel = design->selection_vars.at(ID($abc9_cells));
+
+	for (auto module : design->selected_modules())
+		for (auto cell : module->cells()) {
+			auto inst_module = design->module(cell->type);
+			if (inst_module && inst_module->get_bool_attribute(ID::abc9_flop)) {
+				modules_sel.select(inst_module);
+				// Derive modules for all instantiations of (* abc9_flop *)
+				auto derived_type = inst_module->derive(design, cell->parameters);
+				// And remember one representative cell (for its parameters)
+				if (modules_sel.selected_modules.insert(derived_type).second)
+					cells_sel.select(module, cell);
+			}
+		}
+}
 
-	for (auto cell : module->cells()) {
-		if (cell->type != ID($__ABC9_FF_))
-			continue;
+void prep_dff_map(RTLIL::Design *design)
+{
+	for (auto module : design->modules()) {
+		vector<Cell*> specify_cells;
+		SigBit D, Q;
+		for (auto cell : module->cells())
+			if (cell->type.in(ID($_DFF_N_), ID($_DFF_P_))) {
+				if (D != SigBit())
+					log_error("More than one $_DFF_[NP]_ cell found in module '%s' marked (* abc9_flop *)\n", log_id(module));
+				D = cell->getPort(ID::D);
+				Q = cell->getPort(ID::Q);
+
+				// TODO: Can we avoid doing this?
+				// Convert (* init *) on $_DFF_[NP]_.Q to (* abc9_init *) attr on cell
+				log_assert(GetSize(Q.wire) == 1);
+				auto it = Q.wire->attributes.find(ID::init);
+				Const init;
+				if (it != Q.wire->attributes.end()) {
+					log_assert(GetSize(it->second) == 1);
+					init = it->second;
+					Q.wire->attributes.erase(it);
+				}
+				else
+					init = State::Sx;
+				auto r YS_ATTRIBUTE(unused) = cell->attributes.insert(std::make_pair(ID::abc9_init, init));
+				log_assert(r.second);
+				if (init == State::S1) {
+					log_warning("Module '%s' contains a %s cell with non-zero initial state -- this is not unsupported for ABC9 sequential synthesis. Treating as a blackbox.\n", log_id(module), log_id(cell->type));
```

Comment:
I'm guessing this should be either `not supported` or `unsupported`?

Reply:
Great spot! Resolved in 568572a09.